### ISSUE TITLE
PDP-10 KCC compiler doesn't work

### ISSUE THREAD

**Issue by larsbrinkhoff** [simh/simh#403]
I have installed binaries for the KCC C compiler in ITS.  They work fine with the KLH10 emulator, but not with SIMH.

This is mostly a placeholder until I can provide a good test case.  For now, most information can be found in https://github.com/PDP-10/its/issues/419

SIMH version used is V4.0-0 Beta, git commit id: d8dbc7e6.

**Comment by larsbrinkhoff** [comment]
1. Unpack http://lars.nocrew.org/tmp/kcc.tar.gz
2. Start the emulator with `pdp10 boot`
3. At the `DSKDMP` prompt, type `its`, press *Enter* and then *Escape* *g*.
4. Having seen `SYSTEM JOB USING THIS CONSOLE`, type *Control-Z*.
5. Type `:login lars` and press *Enter*.
6. Type `:c;cc -c test` and press *Enter*.
7. After a few seconds, you will see the error message

       ILOPR; 42>>-25,,757253   757253/   104,,154

   Expected output is:

       5:   INVALID CHARACTER
       5:   INVALID CHARACTER
       5:   INVALID CHARACTER
       
       :KILL
       *

**Comment by larsbrinkhoff** [comment]
Note: I previously thought this was the KCC compiler, but based on new information, it's the one written by Alan Snyder.  I edited some text to reflect this.

**Comment by archivisth** [comment]
I don't have a clue what ITS is doing or even complaining about, but ILOPR is an illegal instruction error. Can someone with some DDT experience comment on what the printout might mean? If you can establish the PC of the failing instruction, put a breakpoint there and run with CPU history set to a reasonable value, like 1000 instructions. That will allow some understanding of what the program was doing.

**Comment by larsbrinkhoff** [comment]
Right, `ILOPR` means illegal operation.  `42>>` indicates the PC of the failing instruction.  `-25,,757253` is the contents at that location.

DDT tries to be helpful by also providing more information about the instruction.  `757253/   104,,154` is another address and its contents.

**Comment by archivisth** [comment]
As I said, setting a breakpoint at 42 and then looking at the history works like a charm. However, I'm unable to interpret what is going on. Here's a dump of the preceding 4000 instruction cycles:

[kcc_history4000.txt](https://github.com/simh/simh/files/1045119/kcc_history4000.txt)

The problem occurs more than 1000 cycles back, where there's a complex series of manipulations of register 17, and the code jumps off into hyperspace:

`014665  000000005640  000001  CONSO PI,1
014667  244321200000  014573  JRST 0,14573
014573  000000000000  104571  WRUBR 104571
014574  000000000040  757124  JRSTF @123025
757124  310000055110  757126  JRST 0,757126
757126  000001000000  000141  HRRZ 3,101(2)
757127  000000000000  000141  HRRM 1,101(2)
757130  000000000000  000003  MOVE 1,3
757131  001530007667  000003  SUBI 17,3
757132  310000055110  056502  JRST 0,@3(17)
056502  001530007664  000003  SUBI 17,3
056503  310000055110  056250  JRST 0,@2(17)
056250  000000057261  000000  MOVEI 1,0
056251  001530007661  000003  SUBI 17,3
056252  310000055110  055137  JRST 0,@2(17)
055137  000000000000  007577  MOVE 1,777721(17)
055140  001530007656  000063  SUBI 17,63
055141  310000055110  000000  JRST 0,@3(17)
000000  310000055110  055110  CAM 0,55110
000001  310000055110  000000  (undefined) 000000000000`

At that point, the code seems  to make its way linearly from 1 to 42, ignoring attempts to execute 0s, and eventually stumbling across an illegal instruction at 42.

**Comment by larsbrinkhoff** [comment]
Thank you, that's more progress than I have been able to make.  (The problem is that the C compiler dynamically maps in code at run time, so it's hard to set breakpoints in advance.)

Running from the accumulator locations 0-17 is unusual, so most likely the jump from 55141 to 0 is wrong.  But it's hard to say why this happens, because it's an indirect jump.

Accumulator 17 is conventionally the stack pointer.  I haven't seen this kind of code pattern in other ITS programs.  I wonder if subtracting something from the stack pointer and then doing an indirect jump through a location on the stack is some kind of subroutine return?  If that's the case, it could be that the stack has been corrupted.  Or the stack pointer has been offset.

**Comment by larsbrinkhoff** [comment]
New developments:

1. There's a third PDP-10 simulator that also runs the C compiler without this problem.
2. I found an actual ITS application written in C, so now there's a real use case for fixing this.

**Comment by larsbrinkhoff** [comment]
I set a breakpoint at 55141.  This is from the KA10 simulator:

```
$1B; 55141>>GO @C(P)   E.A. _ ZEXECV+66
p[   1602007572
/   0
7573/   ZPHSPR+36
7574/   7114
7575/   CAM ZEXECV+66
7576/   0
```

And this is from SIMH's KS10 simulator:

```
$1B; 55141>>GO @C(P)   E.A. _ 0
p[   1603007573
/   ZPHSPR+36
7574/   7114
7575/   CAM ZEXECV+66
7576/   0
```

Apparently, P, the stack pointer, is off by one.  The `GO @C(P)` is `JRST @3(PC)` which is intented to return from a function back to ZEXECV+66.

**Comment by larsbrinkhoff** [comment]
The first problem is that the program uses UFA.  SIMH does the right thing by handling it as an MUUO.  KLH10 apparently implements the instruction even though it's running in KS10 mode.  The KA10 simulator of course has no problem with this instruction.

The program might have been able to go to the UUO handler and return back and proceed, but there's a second problem.  The handler goes to the ILLUUO routine which pushes a word onto the stack.  I'm not sure if this is by design or a bug in the C library.  In any case, when the UUO handler returns, the stack pointer is off by one.

**Comment by markpizz** [comment]
So, your conclusion is that this never actually worked on a KS10 due to a software error in the C runtime?

**Comment by larsbrinkhoff** [comment]
This program probably never worked on a KS10.  I write "probably" because I haven't tested it on a real KS10, but according to documentation it doesn't implement the UFA instruction.  SIMH behaviour faithfully simulates the real hardware.

It's not exactly an error in the runtime, because it was written for KA10 processors which do implement UFA.  By the time ITS was ported to the KS10, the C compiler was no longer in use.
